File: movie_merge/movie_merge/simple_time.rb
class SimpleTime
  attr_reader :hours, :minutes, :seconds, :milliseconds

  def initialize(hours, minutes, seconds, milliseconds = 0)
    initial_seconds = milliseconds / 1000
    initial_minutes = seconds / 60
    initial_hours = minutes / 60

    @hours = hours + initial_hours
    @minutes = (minutes % 60) + initial_minutes
    @seconds = seconds % 60 + initial_seconds
    @milliseconds = milliseconds % 1000
  end

  def to_s
  	@hours.to_s.rjust(2, '0') + ":" +
  	@minutes.to_s.rjust(2, '0') + ":" +
  	@seconds.to_s.rjust(2, '0') + "." +
    @milliseconds.to_s.rjust(3, '0')
  end

  def self.parse(str)
  	parts = str.split(":")

    if parts.size == 3
  	 hours = parts[0].to_i
  	 minutes = parts[1].to_i
  	 seconds = parts[2].to_i
    elsif parts.size == 2
      hours = 0
      minutes = parts[0].to_i
      seconds = parts[1].to_i
    else
      raise "Unrecognized format"
    end
    
  	SimpleTime.new(hours, minutes, seconds)
  end

  def self.from_seconds(in_seconds)
    whole = in_seconds.to_f.truncate
    milliseconds = ((in_seconds.to_f - whole).round(3) * 1000).truncate

    hours = whole / 3600
    minutes = (whole - (3600 * hours)) / 60
    seconds = whole % 60
    SimpleTime.new(hours, minutes, seconds, milliseconds)
  end

  def ==(other)
    if other.nil?
      return false
    elsif !other.instance_of? SimpleTime
      return false
    else
      return (other.hours == self.hours) && (other.minutes == self.minutes) && (other.seconds == self.seconds) && (other.milliseconds == self.milliseconds)
    end
  end

  def +(other)
    if !other.instance_of? SimpleTime
      raise "Cannot add other object types to a SimpleTime"
    else
      return SimpleTime.new(self.hours + other.hours, self.minutes + other.minutes, self.seconds + other.seconds, self.milliseconds + other.milliseconds)
    end
  end
end

File: scala_movie_merge/src/main/scala/com/eaw/moviemerge/SimpleTime.scala
package com.eaw.moviemerge

import com.eaw.moviemerge.DoubleUtils._

/**
  * Created by erikwenberg on 12/24/17.
  */
class SimpleTime(val hours: Int, val minutes: Int, val seconds: Int, val milliseconds: Int = 0) {

  override def toString() = {
    "%02d".format(hours) + ":" +
    "%02d".format(minutes) + ":" +
    "%02d".format(seconds) + "." +
    "%03d".format(milliseconds)
  }

  override def equals(obj: scala.Any): Boolean = {
    if (!obj.isInstanceOf[SimpleTime]) {
      false
    }
    val other = obj.asInstanceOf[SimpleTime]
    if (this.hours == other.hours &&
    this.minutes == other.minutes &&
    this.seconds == other.seconds &&
    this.milliseconds == other.milliseconds) {
      true
    } else {
      false
    }
  }

  def +(other: SimpleTime): SimpleTime = {
    SimpleTime(this.hours + other.hours, this.minutes + other.minutes, this.seconds + other.seconds, this.milliseconds + other.milliseconds)
  }

  def toMilliseconds() = {
    this.milliseconds +
    this.seconds * 1000 +
    this.minutes * 60 * 1000 +
    this.hours * 60 * 60 * 1000
  }
}

object SimpleTime {
  def apply(hours: Int, minutes: Int, seconds: Int, milliseconds: Int = 0) = {
    val initialSeconds = milliseconds / 1000
    val initialMinutes = seconds / 60
    val initialHours = minutes / 60

    new SimpleTime(hours + initialHours, (minutes%60) + initialMinutes, (seconds%60) + initialSeconds, milliseconds % 1000)
  }

  def fromSeconds(s: Double): SimpleTime = {
    val whole = s.toDouble.truncateAt(0)
    val milliseconds = (((s.toDouble - whole).roundAt(3) * 1000).truncateAt(0)).toInt

    val hours = (whole / 3600).toInt
    val minutes = ((whole - (3600 * hours)) / 60).toInt
    val seconds = (whole % 60).toInt
    SimpleTime(hours, minutes, seconds, milliseconds)
  }

  def fromSeconds(s: String): SimpleTime = {
    fromSeconds(s.toDouble)
  }

  def parse(s: String) = {
    def parts = s.split(":")
    if (parts.length == 3) {
      new SimpleTime(parts(0).toInt, parts(1).toInt, parts(2).toInt)
    } else if (parts.length == 2) {
      new SimpleTime(0, parts(0).toInt, parts(1).toInt)
    } else {
      throw new RuntimeException("Unrecognized format")
    }
  }
}
